File: crates/okbayd/src/atlas_ce.rs
//! CE Atlas data bridge — CEData JSON for CuriosityDataSource / KnowledgeAtlas.
//! Slice 0: canvas payload (stub body_html). Endpoint: GET /api/atlas/data
use crate::{graph, paths, theme};
use serde_json::{json, Map, Value};
use std::collections::{HashMap, HashSet};
use std::path::Path;
use std::time::{SystemTime, UNIX_EPOCH};

fn normalize_id(id: &str) -> String {
    let s = id.trim();
    if let Some(stripped) = s.strip_suffix(".md") {
        stripped.to_string()
    } else {
        s.to_string()
    }
}

fn canonical_type(raw: &str, title: &str) -> String {
    let key = raw.trim().to_lowercase().replace(' ', "-");
    let mapped = match key.as_str() {
        "analysis" | "analyses" => Some("analysis"),
        "concept" | "concepts" => Some("concept"),
        "entity" | "entities" => Some("entity"),
        "evidence" => Some("evidence"),
        "fact" | "facts" => Some("fact"),
        "figure" | "figures" => Some("figure"),
        "table" | "tables" | "extracted-table" | "summary-table" => Some("table"),
        "source" | "sources" | "source-note" => Some("source"),
        "note" | "notes" => Some("note"),
        "todo" | "todo-list" => Some("todo-list"),
        "project" | "projects" => Some("project"),
        "hub" => Some("hub"),
        "missing" => Some("missing"),
        "unclassified" => Some("unclassified"),
        _ => None,
    };
    if let Some(m) = mapped {
        if m != "unclassified" {
            return m.to_string();
        }
    }
    // Title prefix backfill: "[con] Foo"
    let t = title.trim();
    if let Some(rest) = t.strip_prefix('[') {
        if let Some(end) = rest.find(']') {
            let stem = rest[..end].trim().to_lowercase();
            let back = match stem.as_str() {
                "con" => Some("concept"),
                "ent" => Some("entity"),
                "ana" => Some("analysis"),
                "src" => Some("source"),
                "evi" => Some("evidence"),
                "fact" => Some("fact"),
                "tbl" | "tab" => Some("table"),
                "fig" => Some("figure"),
                "note" => Some("note"),
                "todo" => Some("todo-list"),
                "proj" => Some("project"),
                _ => None,
            };
            if let Some(b) = back {
                return b.to_string();
            }
        }
    }
    if key == "hub" || key == "missing" {
        return key;
    }
    mapped.unwrap_or("unclassified").to_string()
}

fn is_file_id(id: &str) -> bool {
    id.starts_with("file:")
}

fn iso_now_or_mtime(ws: &Path) -> String {
    let path = paths::graph_json(ws);
    if let Ok(meta) = std::fs::metadata(&path) {
        if let Ok(mtime) = meta.modified() {
            if let Ok(dur) = mtime.duration_since(UNIX_EPOCH) {
                return format_iso(dur.as_secs());
            }
        }
    }
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs())
        .unwrap_or(0);
    format_iso(now)
}

fn format_iso(secs: u64) -> String {
    // Minimal UTC ISO8601 without chrono dependency.
    const SECS_PER_DAY: u64 = 86400;
    const DAYS_PER_CYCLE: u64 = 146097; // 400y
    const SECS_PER_HOUR: u64 = 3600;
    let days = secs / SECS_PER_DAY;
    let tod = secs % SECS_PER_DAY;
    let hh = tod / SECS_PER_HOUR;
    let mm = (tod % SECS_PER_HOUR) / 60;
    let ss = tod % 60;
    // Civil from days since 1970-01-01 (Howard Hinnant algorithm).
    let z = days + 719468;
    let era = z / DAYS_PER_CYCLE;
    let doe = z - era * DAYS_PER_CYCLE;
    let yoe = (doe - doe / 1460 + doe / 36524 - doe / 146096) / 365;
    let y = (yoe as i64) + (era as i64) * 400;
    let doy = doe - (365 * yoe + yoe / 4 - yoe / 100);
    let mp = (5 * doy + 2) / 153;
    let d = doy - (153 * mp + 2) / 5 + 1;
    let m = if mp < 10 { mp + 3 } else { mp - 9 };
    let y = if m <= 2 { y + 1 } else { y };
    format!("{:04}-{:02}-{:02}T{:02}:{:02}:{:02}Z", y, m, d, hh, mm, ss)
}

/// Build CEData-compatible JSON from the okbay graph + theme palette.
pub fn build_ce_data(ws: &Path) -> Value {
    let g = graph::load(ws);
    let theme_val = theme::atlas_theme();
    let palette = theme_val
        .get("types")
        .or_else(|| theme_val.get("kinds"))
        .cloned()
        .unwrap_or_else(|| json!({}));
    // Prefer graph kind (wiki parse folds frontmatter type:→kind on rebuild).
    // Avoid re-listing the whole wiki on every /api/atlas/data hit (Biocure ~40k).
    let mut degree: HashMap<String, i64> = HashMap::new();
    let mut edges_out: Vec<Value> = Vec::new();
    for e in &g.edges {
        let src = normalize_id(&e.source);
        let tgt = normalize_id(&e.target);
        if src.is_empty() || tgt.is_empty() || is_file_id(&src) || is_file_id(&tgt) {
            continue;
        }
        edges_out.push(json!({"source": src, "target": tgt, "type": e.kind}));
        *degree.entry(normalize_id(&e.source)).or_insert(0) += 1;
        *degree.entry(normalize_id(&e.target)).or_insert(0) += 1;
    }

    let mut nodes_out: Vec<Value> = Vec::new();
    let mut pages_out = Map::new();
    let mut seen = HashSet::new();
    for n in &g.nodes {
        let nid = normalize_id(&n.id);
        if nid.is_empty() || is_file_id(&nid) || !seen.insert(nid.clone()) {
            continue;
        }
        let title = if n.title.is_empty() {
            nid.clone()
        } else {
            n.title.clone()
        };
        let ntype = canonical_type(&n.kind, &title);
        let deg = *degree.get(&nid).unwrap_or(&0);
        nodes_out.push(json!({
            "id": nid,
            "path": n.path,
            "type": ntype,
            "title": title,
            "degree": deg,
        }));
        pages_out.insert(
            nid.clone(),
            json!({
                "id": nid,
                "title": title,
                "type": ntype,
                "path": n.path,
                "properties": {"sources": n.sources, "files": n.files},
                "sources": n.sources,
                "files": n.files,
                "body_html": "",
            }),
        );
    }

    json!({
        "workspace": ws.display().to_string(),
        "generated_at": iso_now_or_mtime(ws),
        "palette": palette,
        "nodes": nodes_out,
        "edges": edges_out,
        "pages": pages_out,
        "page_count": g.pages,
    })
}

File: src/okbay/atlas_ce.py
"""CE Atlas data bridge — build CuriosityDataSource-compatible CEData from okbay graph.

Slice 0: canvas payload only (stub body_html). See docs/CE-ATLAS-PARITY-AUDIT.md §3.3 / §7.
Endpoint: GET /api/atlas/data
"""
from __future__ import annotations

import json
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from . import graph as graph_mod
from . import paths, theme, wiki

# Mirror knowledge-atlas datasources/curiosity.ts TYPE_CANONICAL (singular forms).
_TYPE_CANONICAL = {
    "analysis": "analysis",
    "analyses": "analysis",
    "concept": "concept",
    "concepts": "concept",
    "entity": "entity",
    "entities": "entity",
    "evidence": "evidence",
    "fact": "fact",
    "facts": "fact",
    "figure": "figure",
    "figures": "figure",
    "table": "table",
    "tables": "table",
    "extracted-table": "table",
    "summary-table": "table",
    "source": "source",
    "sources": "source",
    "note": "note",
    "notes": "note",
    "todo": "todo-list",
    "todo-list": "todo-list",
    "project": "project",
    "projects": "project",
    "procedure": "procedure",
    "procedures": "procedure",
    "execution": "execution",
    "executions": "execution",
    "hub": "hub",
    "missing": "missing",
    "unclassified": "unclassified",
    "source-note": "source",
}

_PREFIX_TO_TYPE = {
    "con": "concept",
    "ent": "entity",
    "ana": "analysis",
    "src": "source",
    "evi": "evidence",
    "fact": "fact",
    "tbl": "table",
    "tab": "table",
    "fig": "figure",
    "note": "note",
    "todo": "todo-list",
    "proj": "project",
    "proc": "procedure",
    "exec": "execution",
}


def normalize_id(id_or_path: str) -> str:
    s = (id_or_path or "").strip()
    if s.endswith(".md"):
        return s[:-3]
    return s


def canonical_type(raw: str, title: str | None = None) -> str:
    key = (raw or "").strip().lower().replace(" ", "-")
    mapped = _TYPE_CANONICAL.get(key)
    if mapped and mapped != "unclassified":
        return mapped
    if title:
        # Title prefix like "[con] Foo" → concept (CE Switchbay backfill).
        t = title.strip()
        if t.startswith("[") and "]" in t:
            stem = t[1 : t.index("]")].strip().lower()
            back = _PREFIX_TO_TYPE.get(stem)
            if back:
                return back
    return mapped or (key if key in ("hub", "missing") else "unclassified")


def _is_file_id(nid: str) -> bool:
    return (nid or "").startswith("file:")


def _edge_endpoints(edge: dict) -> tuple[str, str]:
    src, tgt = edge.get("source"), edge.get("target")
    if isinstance(src, dict):
        src = src.get("id")
    if isinstance(tgt, dict):
        tgt = tgt.get("id")
    return normalize_id(str(src or "")), normalize_id(str(tgt or ""))


def _wiki_type_overrides(ws: Path | None) -> dict[str, str]:
    """Prefer CE frontmatter `type:` over stored kind when present."""
    overrides: dict[str, str] = {}
    try:
        for page in wiki.list_pages(paths.wiki(ws)):
            front = page.front or {}
            raw = front.get("type") or front.get("kind") or page.kind
            if raw:
                overrides[page.stem] = canonical_type(str(raw), page.title)
    except OSError:
        pass
    return overrides


def build_ce_data(
    graph: dict | None = None,
    ws: Path | None = None,
    palette: dict[str, str] | None = None,
    *,
    enrich_wiki_types: bool = False,
) -> dict[str, Any]:
    """Transform okbay `/graph` JSON into CEData for CuriosityDataSource.

    Uses node ``kind`` (canonicalised) by default. Pass ``enrich_wiki_types=True``
    to re-scan wiki frontmatter ``type:`` (costly on Biocure); prefer rebuilding
    the graph after the wiki ``type:``→kind parse instead.
    """
    root = ws or paths.workspace()
    g = graph if graph is not None else graph_mod.load(root)
    palette = palette if palette is not None else theme.type_palette()
    type_by_stem = _wiki_type_overrides(root) if enrich_wiki_types else {}

    raw_edges = g.get("edges") or []
    edges: list[dict[str, str]] = []
    degree: dict[str, int] = {}
    for e in raw_edges:
        src, tgt = _edge_endpoints(e)
        if not src or not tgt or _is_file_id(src) or _is_file_id(tgt):
            continue
        etype = str(e.get("type") or "wikilink")
        edges.append({"source": src, "target": tgt, "type": etype})
        degree[src] = degree.get(src, 0) + 1
        degree[tgt] = degree.get(tgt, 0) + 1

    nodes_out: list[dict[str, Any]] = []
    pages_out: dict[str, dict[str, Any]] = {}
    seen: set[str] = set()
    for n in g.get("nodes") or []:
        nid = normalize_id(str(n.get("id") or ""))
        if not nid or _is_file_id(nid) or nid in seen:
            continue
        seen.add(nid)
        title = str(n.get("title") or nid)
        raw_kind = str(n.get("kind") or n.get("type") or "note")
        ntype = type_by_stem.get(nid) or canonical_type(raw_kind, title)
        path = str(n.get("path") or "")
        sources = list(n.get("sources") or [])
        files = list(n.get("files") or [])
        deg = int(degree.get(nid, 0))
        nodes_out.append(
            {"id": nid, "path": path, "type": ntype, "title": title, "degree": deg}
        )
        pages_out[nid] = {
            "id": nid,
            "title": title,
            "type": ntype,
            "path": path,
            "properties": {"sources": sources, "files": files},
            "sources": sources,
            "files": files,
            "body_html": "",
        }

    # page_count: prefer integer from graph; else unique non-file nodes.
    page_count = g.get("pages")
    if not isinstance(page_count, int):
        page_count = len(nodes_out)

    generated = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
    graph_path = paths.graph_json(root)
    if graph_path.is_file():
        try:
            mtime = datetime.fromtimestamp(graph_path.stat().st_mtime, tz=timezone.utc)
            generated = mtime.strftime("%Y-%m-%dT%H:%M:%SZ")
        except OSError:
            pass

    return {
        "workspace": str(root),
        "generated_at": generated,
        "palette": dict(palette),
        "nodes": nodes_out,
        "edges": edges,
        "pages": pages_out,
        "page_count": page_count,
    }



def enrich_graph_kinds(ws: Path | None = None) -> dict[str, Any]:
    """One-shot: re-read wiki frontmatter ``type:`` into graph.json node kinds.

    Prefer this (or ``okbay graph rebuild``) over ``enrich_wiki_types=True`` on
    the hot ``/api/atlas/data`` path — Biocure ~40k pages is expensive per request.
    """
    root = ws or paths.workspace()
    g = graph_mod.load(root)
    overrides = _wiki_type_overrides(root)
    changed = 0
    for n in g.get("nodes") or []:
        nid = normalize_id(str(n.get("id") or ""))
        if not nid or nid not in overrides:
            continue
        new_kind = overrides[nid]
        if str(n.get("kind") or "") != new_kind:
            n["kind"] = new_kind
            changed += 1
    out = paths.graph_json(root)
    out.parent.mkdir(parents=True, exist_ok=True)
    out.write_text(json.dumps(g, indent=2), encoding="utf-8")
    return {"ok": True, "changed": changed, "nodes": len(g.get("nodes") or []), "path": str(out)}


def load_ce(ws: Path | None = None) -> dict[str, Any]:
    return build_ce_data(ws=ws)
